packages/langium/src/lsp/language-server.ts

```diff
@@ -36,25 +36,41 @@ export function startLanguageServer(services: LangiumServices): void {
                 documentHighlightProvider: {},
                 codeActionProvider: services.lsp.CodeActionProvider ? {} : undefined,
                 // hoverProvider needs to be created for mouse-over events, etc.
-                hoverProvider: false
+                hoverProvider: false,
+                monikerProvider: {}
             }
         };
+
         if (hasWorkspaceFolderCapability) {
             result.capabilities.workspace = {
                 workspaceFolders: {
                     supported: true
                 }
             };
         }
+
+        if (params.capabilities.workspace?.configuration) {
+            try {
+                // experimental
+                if (params.workspaceFolders)
+                    services.index.IndexManager.initializeWorkspace(params.workspaceFolders);
+            } catch (e) {
+                console.error(e);
+            }
+        }
         return result;
     });
 
+    const langiumDocs = services.documents.LangiumDocuments;
     const documents = services.documents.TextDocuments;
     const documentBuilder = services.documents.DocumentBuilder;
     documents.onDidChangeContent(change => {
-        documentBuilder.build(change.document);
+        documentBuilder.documentChanged(change.document.uri);
+    });
+    documents.onDidClose(() => {
+        if (documents.keys().length === 0)
+            langiumDocs.invalidateAllDocuments(); // clean up all cached document
```

Comment:
Doesn't that just mark them dirty? Cleaning up would mean removing all LangiumDocuments from internal maps etc.

Reply:
I will remove it